draft-how-does-mutant-work.markdown

```diff
@@ -0,0 +1,1630 @@
+
+### Mutator
+
+Ok, so now we have talked about "subjects" a lot. What about the "mutations"?
+
+If you recall the `env` method in `Env::Bootstrap` from above:
+
+```Ruby
+  def env
+    subjects = matched_subjects
+    Env.new(
+      actor_env:        Actor::Env.new(Thread),
+      config:           config,
+      integration:      @integration,
+      matchable_scopes: matchable_scopes,
+      mutations:        subjects.flat_map(&:mutations),
+      parser:           parser,
+      selector:         Selector::Expression.new(@integration),
+      subjects:         subjects
+    )
+  end
+```
+
+you see this line:
+
+```Ruby
+mutations:        subjects.flat_map(&:mutations),
+```
+
+Each `Subject` can yield N mutations, `Mutant expands these in one step here for better progress reports.
+
+Let's check out the [Subject](https://github.com/mbj/mutant/blob/master/lib/mutant/subject.rb) class:
+
+```Ruby
+module Mutant
+  class Subject
+    # Mutations for this subject
+    #
+    # @return [Enumerable<Mutation>]
+    # @return [undefined]
+    def mutations
+      [neutral_mutation].concat(
+        Mutator.mutate(node).map do |mutant|
+          Mutation::Evil.new(self, wrap_node(mutant))
+        end
+      )
+    end
+    memoize :mutations
+  end
+end
+```
+
+This
+
+```Ruby
+Mutator.mutate(node).
+```
+
+is the interesting part into which we're going to dive soon.
+But let me give you a high level overview of what happens here first:
+
+```Ruby
+def mutations
+  [neutral_mutation].concat(
+    Mutator.mutate(node).map do |mutant|
+      Mutation::Evil.new(self, wrap_node(mutant))
+    end
+  )
+end
+```
+
+`Mutator.mutate(node)` takes a given node like this (re-using our example from above):
+
+```
+s(:def, :say_hello,
+  s(:args,         
+  # ...snip
+```
+
+creates multiple mutations out of this node and returns a `Set` of `Parser::AST::Node`.
+We then iterate over this `Set` and create an evil `Mutation` out of each `Parser::AST::Node`. An evil `Mutation` is a mutation for which we expect a test to fail. Or to put it the other way round: If no test fails for this mutation we have a surviving mutant.
+
+Finally, we'll add all of the evil mutations to a "neutral" mutation and all of this makes up the collection of mutations for one specific subject.
+
+Now what's up with that neutral mutation?
+
+A neutral mutation means that you're taking the original node without mutating it:
+
+```Ruby
+def neutral_mutation
+  Mutation::Neutral.new(self, node)
+end
+```
+
+With `Mutation::Neutral` being:
+
+```Ruby
+# Neutral mutation that should not cause mutations to fail tests
+class Neutral < self
+  TEST_PASS_SUCCESS = true
+end
+```
+
+As you can see, in `neutral_mutation` we're basically just passing the original node with `Mutation::Neutral` having the expectation that the test pass on this one.
+The purpose of the neutral mutation is kind of a sanity check to make sure that your original code doesn't break your existing specs.
```

Comment:
Awesome explanation, thanks! I copied your explanation more or less verbatim into the article :)